bundles/admin/admin-layereditor/view/AdminLayerForm/AdminLayerFormHandler.js

```diff
@@ -667,14 +668,41 @@ class UIHandler extends StateHandler {
         });
     }
 
+    /**
+     *
+     * @param {Number} layerId id for layer affected
+     * @param {Object} layer affected layer as WFSlayer object
+     * @param {Object} layerData new fetched layer data
+     */
+    refreshLayerOnMap (layerId, layer, layerData) {
+        if (!layerId || !layer || !layerData) {
+            return;
+        }
+
+        const originalLayerIndex = this.sandbox.getMap().getLayerIndex(layerId); // Save index for the new layer
+        const modifiedLayer = this.mapLayerService.createMapLayer(layerData);
+
+        this.mapLayerService.removeLayer(layerId); // remove existing layer and supress event
+
+        // if layer was found from the selected layers remove and re-add it
+        if (originalLayerIndex !== -1) {
+            this.sandbox.postRequestByName('RemoveMapLayerRequest', [layerId]);
+
+            this.mapLayerService.addLayer(modifiedLayer); // add layer but dont supress event
+            this.sandbox.postRequestByName('AddMapLayerRequest', [layerId]);
+            this.sandbox.postRequestByName('RearrangeSelectedMapLayerRequest', [layerId, originalLayerIndex]);
+        }
```

Comment:
Sending RemoveMapLayerRequest when the layer is not registered on the service might lead to unexpected problems since it triggers an event that can't be fully populated with proper data (== layer details). Also with your code the layer is not added to the service/register either if it was NOT on the map. That needs to be done regardless.

```suggestion
        // if layer was found from the selected layers remove and re-add it
        if (originalLayerIndex !== -1) {
            this.sandbox.postRequestByName('RemoveMapLayerRequest', [layerId]);
        }
        // remove the previous version replace with the new layer data without sending events
        // this is a more secure way of updating all of the layer data for the frontend instead of calling updateLayer that does only partial update
        this.mapLayerService.removeLayer(layerId, true);
        this.mapLayerService.addLayer(modifiedLayer, true);
        // TODO: send maplayer event with "update" operation here

        // if layer was found from the selected layers remove it from map and re-add it
        // this handles everything that needs to be updated on the map without separate code to update and potentially changed data separately
        if (originalLayerIndex !== -1) {
            this.sandbox.postRequestByName('RemoveMapLayerRequest', [layerId]);

            this.sandbox.postRequestByName('AddMapLayerRequest', [layerId]);
            this.sandbox.postRequestByName('RearrangeSelectedMapLayerRequest', [layerId, originalLayerIndex]);
        }
```

Reply:
Sending update event for maplayer is done like this:
```
                // notify components of layer update
                sandbox.notifyAll(Oskari.eventBuilder('MapLayerEvent')(layerId, 'update'));
```

It will probably do less work on most functionalities leading to better performance than having both remove and add events sent.